src/org/alicebot/ab/ProgramABListener.java

```diff
@@ -11,20 +11,20 @@
    * @param predicateName
    * @param result
    */
-  public void onChangePredicate(String predicateName, String result);
+  public void onChangePredicate(Chat chat, String predicateName, String result);
 
   /**
    * Callback when category is added from a the result of matching a learnf tag
    * 
    * @param c - the category added
    */
-  public void onLearnF(Category c);
+  public void onLearnF(Bot bot, Category c);
```

Comment:
we should pass the chat here.. not just the bot.. so you know which user's session triggered the learnf tag.

Reply:
I'm not sure about that - the aiml is "bot" specific ..  predicates are session/chat specific..
Categories are to be turned into aiml, at the moment all aiml is only bot specific.

Unless you want to get fancy and start creating chat/session specific directories for which to save and load subsets of aiml, I don't see the point

Reply:
learnf is global, learn is specific to a session.  We should pass the chat here, not the bot.  you will have access to the bot via the chat.  you will / should have access to the name of the user in the customerId property of the chat (I believe.)

Reply:
sounds good .. I'm for passing more info in the callbacks - so that downstream code can make better decisions ...
at the moment on the mrl ProgramAB callback handler side the learn & learnf are doing the same thing ... so I suspect that will need to be designed differently ...

Reply:
fix pushed